st2api/st2api/controllers/resource.py

```diff
@@ -355,7 +365,28 @@ def _validate_exclude_fields(self, exclude_fields):
         return exclude_fields
 
 
-class ContentPackResourceController(ResourceController):
+class BeseResourceIsolationHandlerMixin(object):
+    """
+    Isolate resources for users except system_user.
+    """
+    @staticmethod
+    def resource_model_filter(model, instances, requester_user=None,
+                              offset=0, eop=0, **kwargs):
+        result = []
+        for instance in instances[offset:eop]:
+            item = model.from_model(instance, **kwargs)
+            if not cfg.CONF.rbac.permission_isolation:
```

Comment:
We should probably also bail out early (first if check, most common scenario) if RBAC is not enabled aka `if not cfg.CONF.rbac.enable`.

Reply:
And perhaps, to make this loop more efficient we could handle those most common scenarios (rbac is not enabled, permission isolation is not enabled) outside the loop - simply fall back to default `resource_model_filter` implementation.

This way we will perform this check for most common scenario only once instead of `N` times where `N` is number of loop iterations.

Reply:
I started working on this and other comments.

There is another thing we should decide - we have a special `observer` system role which grants "view" permission on all the resources.

Should this role also grant "view" on all the resources even when permission isolation is enabled? I think so.

What do others think?

Reply:
@Kami Thank you, I will check if rbac is enabled in the next commit. But I'm not sure I get you that this check only once.